lightning/src/onion_message/offers.rs

```diff
@@ -0,0 +1,96 @@
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Message handling for BOLT 12 Offers.
+
+use core::convert::TryFrom;
+use crate::io::{self, Read};
+use crate::ln::msgs::DecodeError;
+use crate::offers::invoice_error::InvoiceError;
+use crate::offers::invoice_request::InvoiceRequest;
+use crate::offers::invoice::Invoice;
+use crate::offers::parse::ParseError;
+use crate::util::ser::{ReadableArgs, Writeable, Writer};
+
+use crate::prelude::*;
+
+// TLV record types for the `onionmsg_tlv` TLV stream as defined in BOLT 4.
+const INVOICE_REQUEST_TLV_TYPE: u64 = 64;
+const INVOICE_TLV_TYPE: u64 = 66;
+const INVOICE_ERROR_TLV_TYPE: u64 = 68;
+
+/// Possible BOLT 12 Offers messages sent and received via an [`OnionMessage`].
+///
+/// [`OnionMessage`]: crate::ln::msgs::OnionMessage
+#[derive(Debug)]
+pub enum OffersMessage {
+	/// A request for an [`Invoice`] for a particular [`Offer`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	InvoiceRequest(InvoiceRequest),
+
+	/// An [`Invoice`] sent in response to an [`InvoiceRequest`] or a [`Refund`].
+	///
+	/// [`Refund`]: crate::offers::refund::Refund
+	Invoice(Invoice),
+
+	/// An error from handling an [`OffersMessage`].
+	InvoiceError(InvoiceError),
+}
+
+impl OffersMessage {
+	/// Returns whether `tlv_type` corresponds to a TLV record for Offers.
+	pub fn is_known_type(tlv_type: u64) -> bool {
+		match tlv_type {
+			INVOICE_REQUEST_TLV_TYPE | INVOICE_TLV_TYPE | INVOICE_ERROR_TLV_TYPE => true,
+			_ => false,
+		}
+	}
+
+	/// The TLV record type for the message as used in an `onionmsg_tlv` TLV stream.
+	pub fn tlv_type(&self) -> u64 {
+		match self {
+			OffersMessage::InvoiceRequest(_) => INVOICE_REQUEST_TLV_TYPE,
+			OffersMessage::Invoice(_) => INVOICE_TLV_TYPE,
+			OffersMessage::InvoiceError(_) => INVOICE_ERROR_TLV_TYPE,
+		}
+	}
+
+	fn parse(tlv_type: u64, bytes: Vec<u8>) -> Result<Self, ParseError> {
+		match tlv_type {
+			INVOICE_REQUEST_TLV_TYPE => Ok(Self::InvoiceRequest(InvoiceRequest::try_from(bytes)?)),
+			INVOICE_TLV_TYPE => Ok(Self::Invoice(Invoice::try_from(bytes)?)),
+			INVOICE_ERROR_TLV_TYPE => Ok(Self::InvoiceError(InvoiceError::try_from(bytes)?)),
+			_ => unreachable!(),
+		}
+	}
+}
+
+impl Writeable for OffersMessage {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		match self {
+			OffersMessage::InvoiceRequest(message) => message.write(w),
+			OffersMessage::Invoice(message) => message.write(w),
+			OffersMessage::InvoiceError(message) => message.write(w),
+		}
+	}
+}
+
+impl ReadableArgs<u64> for OffersMessage {
+	fn read<R: Read>(r: &mut R, tlv_type: u64) -> Result<Self, DecodeError> {
+		let mut bytes = Vec::new();
+		r.read_to_end(&mut bytes).unwrap();
+
+		match Self::parse(tlv_type, bytes) {
+			Ok(message) => Ok(message),
+			Err(ParseError::Decode(e)) => Err(e),
+			Err(_) => Err(DecodeError::InvalidValue), // TODO: log ParseError?
```

Comment:
Hmm... seems I'd have to update all the `Readable` / `ReadableArgs` impls back to `wire::Message` to take a logger to use here. Perhaps it would be better to make a new `DecodeError` variant wrapping the `ParseError` to avoid that? That was we could log in `PeerManager`. Let me know if you have a preference or another idea.